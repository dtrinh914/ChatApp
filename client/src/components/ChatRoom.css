.ChatRoom{
    height: 95vh;
    width:100%;
}