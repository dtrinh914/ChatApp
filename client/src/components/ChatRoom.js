import React, {useEffect, useContext} from 'react';
import ChatWindow from './ChatWindow';
import ChatInput from './ChatInput';
import './ChatRoom.css';
import io from 'socket.io-client';
import {ChatContext} from '../contexts/chatContext'

let socket;
function ChatRoom(){
    const {chatData, dispatch} = useContext(ChatContext);
    
    useEffect(() => {
        socket = io();
        //listens for new messages from the backend and updates state
        socket.on('message', (room, message) => {
            dispatch({type:'NEWMSG', room:room, message:message})
        });
        
        //on connect joins the rooms on the client side
        socket.on('connect', () => {
            chatData.groups.forEach(group => {
                socket.emit('room', group._id)
            })
        })
    },[])

    const newMessage = (message) => {
        socket.emit('message', chatData.selected, message);
        dispatch({type:'NEW_MSG', room:chatData.selected, message:message});
    }

    return(
        <div className='ChatRoom'>
            <ChatWindow />
            <ChatInput onConfirm={newMessage} />
        </div>
    );
}

export default ChatRoom;