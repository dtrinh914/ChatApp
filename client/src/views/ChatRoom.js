import React, {useState, useEffect} from 'react';
import {useHistory} from 'react-router-dom';
import io from 'socket.io-client';
import ChatWindow from '../components/ChatWindow';
import ChatInput from '../components/ChatInput';
import Navbar from '../components/Navbar';
import './ChatRoom.css';

let socket;
function ChatRoom({username, loggedIn, setUserData}){
    const [messages, setMessages] = useState([]);
    let history = useHistory();

    useEffect(() => {
        //redirect if user is not logged in
        if(!loggedIn){
            history.push('/');
        } else {
            socket = io();
            //listens for new messages from the backend and updates state
            socket.on('chat message', (message) => {
                setMessages(prevMsg => [...prevMsg, message]);
            });
        }
    }, [loggedIn, history]);

    //Sends message to backend
    const newMessage = (message) => {
        socket.emit('chat message', message);
        setMessages(prevMsg => [...prevMsg, message]);
    }

    //Closes all socket connections
    const closeSockets = () => {
        socket.close('chat message');
    }

    return(
        <div>
            <Navbar username={username} history={history} setUserData={setUserData} closeSockets={closeSockets} />
            <ChatWindow messages={messages} />
            <ChatInput newMessage={newMessage} />
        </div>
    );
}

export default ChatRoom;