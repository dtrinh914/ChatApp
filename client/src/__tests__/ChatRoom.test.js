import React from 'react';
import ChatRoom from '../views/ChatRoom';
import { render, cleanup, fireEvent } from '@testing-library/react';

afterEach(cleanup);

jest.mock('react-router-dom', () => ({
    useHistory: () => ({
      push: jest.fn()
    })
}));

test('should match snapshot', () => {
    const {asFragment} = render(<ChatRoom />)
    expect(asFragment()).toMatchSnapshot();
});

test('should type Hello World in chat input, press send, and Hello World should be displayed in chat window', ()=> {
    const {container} = render(<ChatRoom loggedIn={true} />);
    const ChatInputBox = container.querySelector('.ChatInput textarea');
    const ChatInputButton = container.querySelector('.ChatInput button');
    fireEvent.change(ChatInputBox, {target: {value: 'Hello World'}});
    fireEvent.click(ChatInputButton);

    const message = container.querySelector('.ChatWindow li').textContent;
    expect(message).toBe('Hello World');
})