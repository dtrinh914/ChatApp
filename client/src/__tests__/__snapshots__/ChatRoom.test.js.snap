// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="Navbar"
    >
      <h1 />
      <button>
        Log Out
      </button>
    </nav>
    <div
      class="ChatWindow"
    >
      <ul />
    </div>
    <div
      class="ChatInput"
    >
      <textarea
        type="text"
      />
      <button>
        Send
      </button>
    </div>
  </div>
</DocumentFragment>
`;
